components/motor/gpio/loop.go

```diff
@@ -60,27 +75,259 @@ func (m *EncodedMotor) setupControlLoop() error {
+		return err
+	}
+
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	cm.offsetInTicks = -1 * offset * cm.ticksPerRotation
+	return nil
+}
+
+// GoTo instructs the motor to go to a specific position (provided in revolutions from home/zero),
+// at a specific speed. Regardless of the directionality of the RPM this function will move the motor
+// towards the specified target/position
+// This will block until the position has been reached.
+func (cm *controlledMotor) GoTo(ctx context.Context, rpm, targetPosition float64, extra map[string]interface{}) error {
+	// no op manager added, we're relying on GoFor's oepration manager in this driver
+	pos, err := cm.Position(ctx, extra)
+	if err != nil {
+		return err
+	}
+	currRotations := pos / cm.ticksPerRotation
+	rotations := targetPosition - currRotations
+	// if you call GoFor with 0 revolutions, the motor will spin forever. If we are at the target,
+	// we must avoid this by not calling GoFor.
+	if rdkutils.Float64AlmostEqual(rotations, 0, 0.1) {
+		cm.logger.CDebug(ctx, "GoTo distance nearly zero, not moving")
+		return nil
+	}
+	return cm.GoFor(ctx, rpm, rotations, extra)
+}
+
+// GoFor instructs the motor to go in a specific direction for a specific amount of
+// revolutions at a given speed in revolutions per minute. Both the RPM and the revolutions
+// can be assigned negative values to move in a backwards direction. Note: if both are
+// negative the motor will spin in the forward direction.
+// If revolutions is 0, this will run the motor at rpm indefinitely
+// If revolutions != 0, this will block until the number of revolutions has been completed or another operation comes in.
+func (cm *controlledMotor) GoFor(ctx context.Context, rpm, revolutions float64, extra map[string]interface{}) error {
+	cm.opMgr.CancelRunning(ctx)
+	ctx, done := cm.opMgr.New(ctx)
+	defer done()
+
+	switch speed := math.Abs(rpm); {
+	case speed < 0.1:
+		cm.logger.CWarn(ctx, "motor speed is nearly 0 rev_per_min")
+		return motor.NewZeroRPMError()
+	case cm.real.maxRPM > 0 && speed > cm.real.maxRPM-0.1:
+		cm.logger.CWarnf(ctx, "motor speed is nearly the max rev_per_min (%f)", cm.real.maxRPM)
+	default:
+	}
+
+	currentTicks, _, err := cm.enc.Position(ctx, encoder.PositionTypeTicks, extra)
+	if err != nil {
+		return err
+	}
+
+	goalPos, _, _ := encodedGoForMath(rpm, revolutions, currentTicks, cm.ticksPerRotation)
+
+	if cm.loop == nil {
+		// create new control loop if control config exists
+		if len(cm.controlLoopConfig.Blocks) != 0 {
+			if err := cm.startControlLoop(); err != nil {
+				return err
+			}
+		}
+	}
+
+	cm.loop.Resume()
+	// set control loop values
+	velVal := math.Abs(rpm * cm.ticksPerRotation / 60)
+	// when rev = 0, only velocity is controlled
+	// setPoint is +/- infinity, maxVel is calculated velVal
+	if err := cm.updateControlBlock(ctx, goalPos, velVal); err != nil {
+		return err
+	}
+
+	// we can probably use something in controls to make GoFor blockign without this
```

Comment:
I worry that if we make that change, it will cause issues for other apis. imo it seems better to have the control loop never be blocking and block from the APIs, because how will we know if the control loop is being called from (for example) setVelocity or goFor? maybe there's a way to figure out what api is starting the control loop with some extra flag but I haven't thought about it too hard

Reply:
I was mor ethinkg of using a function that accesses some boolean state rather than making the control loop itself blocking, like `IsRunning` and special casing the speed only set on GoFor at this level, but I won't do it in this pr. 

I can remove the comment if you like.

Reply:
ah I see yes you probably can use running great point

Reply:
I think we still need some kind of condition/exit check, since the only things that set running to false are stop & pause. using the helper function + WaitForSuccess seems like a reasonable way to have the signal be blocking